// @workadventure/tiled-map-type-guard/dist/ITiledMapObject.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTiledMapObject = void 0;
const tg = __importStar(require("generic-type-guard"));
const ITiledMapProperty_1 = require("./ITiledMapProperty");
const ITiledMapPoint_1 = require("./ITiledMapPoint");
const ITiledMapText_1 = require("./ITiledMapText");
exports.isTiledMapObject = new tg.IsInterface()
    .withProperties({
    id: tg.isNumber,
    name: tg.isString,
    type: tg.isString,
    visible: tg.isBoolean,
    x: tg.isNumber,
    y: tg.isNumber,
})
    .withOptionalProperties({
    ellipse: tg.isBoolean,
    gid: tg.isBoolean,
    height: tg.isNumber,
    point: tg.isBoolean,
    polygon: tg.isArray(ITiledMapPoint_1.isTiledMapPoint),
    polyline: tg.isArray(ITiledMapPoint_1.isTiledMapPoint),
    properties: tg.isArray(ITiledMapProperty_1.isTiledMapProperty),
    rotation: tg.isNumber,
    template: tg.isString,
    text: ITiledMapText_1.isTiledMapText,
    width: tg.isNumber,
})
    .get();
