// @workadventure/tiled-map-type-guard/dist/ITiledMapGroupLayer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTiledMapGroupLayer = exports.setIsTiledMapLayer = void 0;
const tg = __importStar(require("generic-type-guard"));
const ITiledMapProperty_1 = require("./ITiledMapProperty");
let circularDependencyBreaker = tg.isNumber;
const setIsTiledMapLayer = (newFunction) => {
    circularDependencyBreaker = newFunction;
};
exports.setIsTiledMapLayer = setIsTiledMapLayer;
exports.isTiledMapGroupLayer = new tg.IsInterface()
    .withProperties({
    name: tg.isString,
    opacity: tg.isNumber,
    type: tg.isSingletonString('group'),
    layers: tg.isArray((param) => circularDependencyBreaker(param)),
    visible: tg.isBoolean,
})
    .withOptionalProperties({
    height: tg.isNumber,
    draworder: tg.isString,
    id: tg.isNumber,
    offsetx: tg.isNumber,
    offsety: tg.isNumber,
    parallaxx: tg.isNumber,
    parallaxy: tg.isNumber,
    properties: tg.isArray(ITiledMapProperty_1.isTiledMapProperty),
    startx: tg.isNumber,
    starty: tg.isNumber,
    tintcolor: tg.isString,
    width: tg.isNumber,
    x: tg.isNumber,
    y: tg.isNumber,
})
    .get();
