// @workadventure/tiled-map-type-guard/dist/ITiledMapTile.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTiledMapTile = void 0;
const tg = __importStar(require("generic-type-guard"));
const ITiledMapFrame_1 = require("./ITiledMapFrame");
const ITiledMapObjectLayer_1 = require("./ITiledMapObjectLayer");
const ITiledMapProperty_1 = require("./ITiledMapProperty");
const ITiledMapTerrain_1 = require("./ITiledMapTerrain");
exports.isTiledMapTile = new tg.IsInterface()
    .withProperties({
    id: tg.isNumber,
})
    .withOptionalProperties({
    animation: tg.isArray(ITiledMapFrame_1.isTiledMapFrame),
    image: tg.isString,
    imageheight: tg.isNumber,
    imagewidth: tg.isNumber,
    objectgroup: tg.isArray(ITiledMapObjectLayer_1.isTiledMapObjectLayer),
    probability: tg.isNumber,
    properties: tg.isArray(ITiledMapProperty_1.isTiledMapProperty),
    terrain: tg.isArray(ITiledMapTerrain_1.isTiledMapTerrain),
    type: tg.isString,
})
    .get();
